### DISCUSSION TITLE
Switch DirichletBC to NeumannBC based on a calculation in the Materials block

### DISCUSSION THREAD

**Discussion by styyokuda** [idaholab/moose#26522] [Q&A General]
All,
I would like to switch DirichletBC to NeumannBC based on a calculation in the [Materials] block.

For instance, the run gets started with DirichletBC and when the value of sum reaches to 100 in the MaterialPropertyOld class (please see below), I would like to switch DirichletBC to NeumannBC.

I have been looking at the Controls apps: AddControlAction, BoolFunctionControl, ConditionalFunctionEnableControl, LibtorchNeuralNetControl, PIDTransientControl, RealFunctionControl, and TimePeriod, but I have not found the way to do it.
Please help. S. Thomas
```
MaterialPropertyOld::initQpStatefulProperties()
{ _sum[_qp] = _initial_sum;}

MaterialPropertyOld::computeQpProperties()
{_sum[_qp] = _sum_old[_qp] + 10;}
```

**Comment by GiudGiud** [comment]
Hello

So do you want this change to be local (on a per Qp basis, depending on the value of the sum) or global (the whole sideset switches)?

if global:
Postprocessor to report on the material property
ParsedFunction depending on the postprocessor
ConditionalFunctionEnableControl on this function

Guillaume

**Reply by styyokuda** [reply]
Thank you, Guillaume.
Basically, what I would like to do is that the run gets started with the "zero" DirichletBC, and then, switch to the NeumannBC with the flux value of  "sum" (I have modified the NeumannBC class to read the "sum" value) when the "sum" value reaches to 100, for instance.
I am hoping I can do that ... S. Thomas

**Reply by GiudGiud** [reply]
Use the Controls system to make the switch then?
The sum is in a material property right?

**Reply by styyokuda** [reply]
Thank you, Guillaume, as always.
Yes, I read "sum" by _sum(getMaterialProperty<real>("sum")).
I am trying use the Controls system to switch but I cannot find the way to do.
Please help. S. Thomas

**Reply by GiudGiud** [reply]
Use a Postprcoessor to get the sum
Use a ParsedFunction depending on the postprocessor (put the pp in the expression)
then use ConditionalFunctionEnableControl with this function to act on both boundary conditions

**Reply by styyokuda** [reply]
Thank you, Guillaume.
I am still learning your suggestion. Meanwhile I would like to make sure my understanding below.
The condition to switch DirichletBC to NeumannBC in my actual model is more complicated than the "sum" value reaches "100" as I stated above.

I would like to ask that, in the expression below, DirichletBC can be switched to NeumannBC when _bc_switch[_qp] = 1 and use "sum" of _sum(getMaterialProperty("sum")) as the flux value as far as bc_switch is used as _bc_switch(getMaterialProperty("bc_switch")) and your suggestion can be still used.
Please help, S. Thomas
```
MaterialPropertyOld::initQpStatefulProperties()
{ _sum[_qp] = _initial_sum;}

MaterialPropertyOld::computeQpProperties()
{
_sum[_qp] = _sum_old[_qp] + 10;

_bc_switch[_qp] = 0;

If (_sum[_qp] > 100)
   {
    _bc_switch[_qp] = 1;
   }
}
```

**Reply by GiudGiud** [reply]
bc_swtich is a material property. So what I said (use a PP, then parsedFunction, then the COnditionalFunctionEnableControl) applies?

Btw you will not be able to use a single BC object to act both as a Dirichlet and Neumann BC. You ll need to use two objects

**Reply by styyokuda** [reply]
For
MaterialPropertyOld::computeQpProperties()
{_sum[_qp] = _sum_old[_qp] + 10;
_bc_switch[_qp] = 0;
If (_sum[_qp] > 100)
   {_bc_switch[_qp] = 1;}
}
[Controls]
  [./switch_bcs]
    type = ConditionalFunctionEnableControl
    conditional_function = 'conditional_function01'
    enable_objects = 'BCs::NeumannBC'
    disable_objects = 'BCs::DirichletBC'
    execute_on = 'initial timestep_begin timestep_end'
  [../]
[]
[Functions]
  [conditional_function01]
    type = ParsedFunction
    expression = 'bc_switch_ck > 0'
    symbol_names = 'bc_switch_ck'
    symbol_values = 'pp_bc_switch'
  []
[]
[Postprocessors]
  [pp_bc_switch]
    type = PointValue
    point = '0.5 0.5 0.0'
    variable = 'bc_switch'
    execute_on = 'initial timestep_begin timestep_end'
  []
[]

I am getting pp_bc_switch = 0.5, not 1, on the monitor. I don't know why. 
Please help. S. Thomas

**Reply by GiudGiud** [reply]
We re missing the BC block here I think to see what s going on.

And I m correct in assuming the PointValue is being sampled near the boundary?

**Reply by styyokuda** [reply]
Thank you, Guillaume.
Switching DirichletBC to NeumannBC (I am giving the value of _sum[_qp]) looks good. 
Yes, the location of PointValue is near or at the boundary."
"std::cout << _bc_switch[_qp]" prints exactly "1" but pp_bc_switch prints "0.5." They should be the same, shouldn't them?

**Reply by GiudGiud** [reply]
Maybe. I d like to see your variables block with bc_switch and your boundary conditions block to be more sure

**Reply by styyokuda** [reply]
Thank you, Guillaume.
The simplified version of my model input is below (four variables for the diffusion model).
I do appreciate your time and help.
S. Thomas
```
[Mesh]
  [./cartesian]
    type = CartesianMeshGenerator
    dim = 2
    dx = '0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1'
    ix = '10 10 10 10 10 10 10 10 10 10'
    dy = '0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1
          0.1'
    iy = '10 10 10 10 10 10 10 10 10 10'
    subdomain_id = '1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2
                    1 1 1 1 1 2 2 2 2 2'
  [../]

  [./sideset_12]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = 1
    paired_block = 2
    new_boundary = '1_right'
    input = cartesian
  [../]
  [./sideset_21]
    type = SideSetsBetweenSubdomainsGenerator
    primary_block = 2
    paired_block = 1
    new_boundary = '2_left'
    input = sideset_12
  [../]

[]
[Variables]
  [./Temp]
    order = FIRST
    family = LAGRANGE
    block = '1 2'
  [../]
  [./A]
    order = FIRST
    family = LAGRANGE
    block = '1'
  [../]
  [./B]
    order = FIRST
    family = LAGRANGE
    block = '1'
  [../]
  [./C]
    order = FIRST
    family = LAGRANGE
    block = '2'
  [../]
[]

[Kernels]
  [./A_t] 
    type = TimeDerivative
    variable = A
    block = 1
  [../]
  [./A_diff]
    type = MatDiffusion
    variable = A
    diffusivity = 0.3
    block = 1
  [../]

  [./B_t] 
    type = TimeDerivative
    variable = B
    block = 1
  [../]
  [./B_diff]
    type = MatDiffusion
    variable = B
    diffusivity = 0.4
    block = 1
  [../]

  [./C_t]
    type = TimeDerivative
    variable = C
    block = 2
  [../]
  [./C_diff]
    type = MatDiffusion
    variable = C
    diffusivity = 0.5
    block = 2
  [../]

  [./Temp_t]
    type = TimeDerivative
    variable = Temp
    block = '1 2'
  [../]
  [./Temp_diff]
    type = MatDiffusion
    variable = Temp
    diffusivity = 1.0
    block = '1 2'
  [../]
[]

[Controls]
[./switch_bcs]
type = ConditionalFunctionEnableControl
conditional_function = 'conditional_function01'
enable_objects = 'BCs::B_Neumann'
disable_objects = 'BCs::B_Dirichlet'
execute_on = 'initial timestep_begin timestep_end'
[../]
[]

[BCs]
  [./Temp_BC]
    type = DirichletBC
    boundary = 'right top bottom left'
    variable = Temp
    value = 100
  [../]

  [./A_BC]
    type = DirichletBC
    variable = A
    boundary = 'left'
    value = 0.0
  [../]
  [./A_interface]
    type = NeumannValueBC
    variable = A
    boundary = '1_right'
  [../]

  [./C_BC]
    type = DirichletBC
    variable = C
    boundary = 'right'
    value = 0.0
  [../]
  [./C_interface]
    type = NeumannValueBC
    variable = C
    boundary = 2_left
  [../]

  [./B_BC]
    type = DirichletBC
    variable = B
    boundary = 'left'
    value = 1.0 
  [../]
  [./B_Dirichlet]
    type = DirichletBC
    variable = B
    boundary = '1_right'
    value = 0.0
  [../]
  [./B_Neumann]
    type = NeumannValueBC
    variable = B
    boundary = '1_right'
  [../]
[]

[Materials]
  [./material_calculation]
    type = MaterialCalculation
    diffusion_variable = B
    heat_variable= Temp
    block = '1'
  [../]
[]

[Functions]
  [conditional_function01]
     type = ParsedFunction
     expression = 'bc_switch_ck > 0'
     symbol_names = 'bc_switch_ck'
     symbol_values = 'pp_bc_switch'
[]

[Postprocessors]
  [pp_bc_switch]
    type = PointValue
    point = '  0.5 0.5 0.0'
    variable = 'bc_switch'
    execute_on = 'initial timestep_begin timestep_end'
  []
[]

[VectorPostprocessors]
  [oxidation_thickness]
    type = LineMaterialRealSampler
    start = '0.49 0.0 0.0'
    end = '  0.5 0.5 0.0'
    property = 'sum'
    sort_by = y
  []
[]

[Preconditioning]
  [./smp]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type'
    petsc_options_value = 'lu'
  [../]
[]

[Executioner]
  type = Transient
  petsc_options = '-snes_converged_reason'
  line_search = l2
  end_time = 10
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-6
[]

[Debug]
  show_var_residual_norms = true
[]

[Outputs]
    print_linear_residuals = true
    print_nonlinear_residuals = true
  [./console]
    type = Console
    output_file = false
  [../]
  [./out]
    type = Exodus
    interval = 1
    output_material_properties = true
  [../]
  [./csv]
    type = CSV
    file_base = 'sum_cal/sum_cal'
    interval = 1
    time_data = true
  [../]
[../]
```

**Reply by styyokuda** [reply]
[VectorPostprocessors] in the input file I posted previously had an error. The ones below should be it.
Thank you, S Thomas

```
[VectorPostprocessors]
[sum_cal]
type = LineMaterialRealSampler
start = '0.49 0.0 0.0'
end = ' 0.5 0.5 0.0'
property = 'sum'
sort_by = y
[]
[]

[Outputs]
[./csv]
type = CSV
file_base = 'sum_cal/sum_cal'
interval = 1
time_data = true
[../]
[]
```

**Reply by GiudGiud** [reply]
Hello

I cannot run this input, I dont have the MaterialCalculation object. 
There is also no variable called bc_switch anywhere in the input, so I cannot isolate this variable to examine it. 

That would be what I look at to check what the postprocessor should return

Guillaume

**Reply by GiudGiud** [reply]
if bc_switch is a linear lagrange variable, this 0.5 value could just be in betwee a region where it is 1 and a region where it is 0 (as the variable varies linearly)

**Reply by styyokuda** [reply]
Thank you, Guillaume.
I have to simplify the program before I send it to you for your run. Meanwhile, I would like to let you know that bc_switch is defined in the MaterialCalculation object as
```
 _sum(declareProperty<Real>("sum")),
 _bc_switch(declareProperty<Real>("bc_switch")),
```
and assign the value of "0" or "1" to it as;
```
{_sum[_qp] = _sum_old[_qp] + 10;
_bc_switch[_qp] = 0;
If (_sum[_qp] > 100)
{_bc_switch[_qp] = 1;}
```
Then,
```
[Functions]
  [conditional_function01]
     type = ParsedFunction
     expression = 'bc_switch_ck > 0'
     symbol_names = 'bc_switch_ck'
     symbol_values = 'pp_bc_switch'
[]

[Postprocessors]
  [pp_bc_switch]
    type = PointValue
    point = '  0.5 0.5 0.0'
    variable = 'bc_switch'
    execute_on = 'initial timestep_begin timestep_end'
  []
[]

[VectorPostprocessors]
  [oxidation_thickness]
    type = LineMaterialRealSampler
    start = '0.49 0.0 0.0'
    end = '  0.5 0.5 0.0'
    property = 'sum'
    sort_by = y
  []
[]
```
Everything is working correctly except pp_bc_switch prints "0.5" instead "1."
I would like to thank you for your time and effort.
S. Thomas

**Reply by GiudGiud** [reply]
you have bc_switch as a material property first
but the PointValue refers to a bc_switch variable. So what sets this variable? The material property output to a bc_switch aux variable?

**Reply by styyokuda** [reply]
Thank you, Guillaume.
I am trying to use "bc_switch" in the exactly the same way as "sum" so that the other objects can read it; namely, the PointValue for  "bc_switch" and "LineMaterialRealSampler" for "sum."  So,  I don't set up "bc_switch" and  "sum" in the variables block and the aux variable block in the input file. Is this a problem(s)?

**Reply by GiudGiud** [reply]
PointValue does not accept a 'bc_switch' material property
See https://mooseframework.inl.gov/source/postprocessors/PointValue.html
It only accepts variables

so it s not acting on the material property, it s acting on an auxiliary variable which is output for this material property.
This is why you do not have a bc_switch variable in the input file, but it still can be used as a variable in this postprocessor

If you sample this variable in a transition region between 0 and 1 for the material property, I understand why bc_switch is 0.5

**Reply by styyokuda** [reply]
Thank you, Guillaume.
As far as it meets pp_bc_switch (bc_switch) > 0, it works for me. So, I would like to keep using it.
Thank you very much for your time and help.
S. Thomas
